pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,615 @@
+                                                        .NonRecoverableLedgerException) {
+                                                    if (((ManagedLedgerImpl)topic.getManagedLedger())
+                                                            .ledgerExists(index.getAbortedMarkLedgerID())) {
+                                                        log.error("[{}] Failed to read snapshot segment [{}:{}]",
+                                                                topic.getName(), index.segmentLedgerID,
+                                                                index.segmentEntryID, exception);
+                                                        topic.close();
+                                                        handleSegmentFuture.completeExceptionally(exception);
+                                                    } else {
+                                                        indexes.remove(new PositionImpl(
+                                                                index.getAbortedMarkLedgerID(),
+                                                                index.getAbortedMarkEntryID()));
+                                                        hasInvalidIndex.set(true);
+                                                    }
+                                                }
+                                            }
+                                        }, null);
+                            });
+                            openManagedLedgerFuture.complete(null);
+                        }
+
+                        @Override
+                        public void openReadOnlyManagedLedgerFailed(ManagedLedgerException exception, Object ctx) {
+                            log.error("[{}] Failed to open readOnly managed ledger", topic, exception);
+                            openManagedLedgerFuture.completeExceptionally(exception);
+                        }
+                    };
+
+                    TopicName snapshotIndexTopicName = TopicName.get(TopicDomain.persistent.toString(),
+                            TopicName.get(topic.getName()).getNamespaceObject(),
+                            SystemTopicNames.TRANSACTION_BUFFER_SNAPSHOT_SEGMENTS);
+                    this.topic.getBrokerService().getPulsar().getManagedLedgerFactory()
+                            .asyncOpenReadOnlyManagedLedger(snapshotIndexTopicName
+                                            .getPersistenceNamingEncoding(), callback,
+                                    topic.getManagedLedger().getConfig(),
+                                    null);
+                    //Wait the processor recover completely and the allow TB to recover the messages
+                    // after the startReadCursorPosition.
+
+                    return openManagedLedgerFuture
+                            .thenCompose((ignore) -> FutureUtil.waitForAll(completableFutures).thenCompose((i) -> {
```

Comment:
```suggestion
                            .thenCompose((ignore) -> FutureUtil.waitForAll(completableFutures)).thenCompose((i) -> {
```
otherwise we can't catch the openManagedLedgerFuture exceptio